File: gostdc.c
#include "gostdc.h"

int errno;

void
free(void *ptr) {
	realloc(ptr, 0);
}

void *
malloc(size_t size) {
	return realloc(0, size);
}


/*
 * Below this point only stubs/bridges between C and Go should be found.
 */

int
creat(const char *path, uint16 mode) {
	void ·go_creat(uintptr, uintptr, uintptr);
	int32 ret;
	·go_creat((uintptr)path, (uintptr)mode, (uintptr)&ret);
	return ret;
}

int
open(const char *path, int oflag) {
	void ·go_open(uintptr, uintptr, uintptr);
	int32 ret;
	·go_open((uintptr)path, (uintptr)oflag, (uintptr)&ret);
	return ret;
}

int
close(int filedes) {
	void ·go_close(uintptr, uintptr);
	int32 ret;
	·go_close((uintptr)filedes, (uintptr)&ret);
	return ret;
}

int64
lseek(int filedes, int64 offset, int whence) {
	void ·go_lseek(uintptr, uintptr, uintptr, uintptr);
	int64 ret;
	·go_lseek((uintptr)filedes, (uintptr)&offset, (uintptr)whence, (uintptr)&ret);
	return ret;
}

ssize_t
read(int filedes, void *buf, size_t nbyte) {
	void ·go_read(uintptr, uintptr, uintptr, uintptr);
	int32 ret;
	·go_read((uintptr)filedes, (uintptr)buf, (uintptr)nbyte, (uintptr)&ret);
	return ret;
}

ssize_t
write(int filedes, const void *buf, size_t nbyte) {
	void ·go_write(uintptr, uintptr, uintptr, uintptr);
	int32 ret;
	·go_write((uintptr)filedes, (uintptr)buf, (uintptr)nbyte, (uintptr)&ret);
	return ret;
}

void
_exit(int code) {
	void ·go__exit(uintptr);
	·go__exit((uintptr)code);
}

void
abort(void) {
	void ·go_abort(void);
	·go_abort();
}

void*
realloc(void* ptr, size_t size) {
	void ·go_realloc(uintptr, uintptr, uintptr);
	uintptr ret = 0;
	·go_realloc((uintptr)ptr, (uintptr)size, (uintptr)&ret);
	return (void*)ret;
}

clock_t
clock(void)
{
	void ·go_clock(uintptr);
	int32 secs;
	·go_clock((uintptr)&secs);
	return ((long)secs)*((long)CLOCKS_PER_SEC);
}

File: gostdc.go
package gostdc

import (
	"os"
	"time"
	"unsafe"
)

func peek(addr uintptr) *byte {
	return (*byte)(unsafe.Pointer(addr))
}

func go_creat(uintptr, uintptr, uintptr)          { panic("nyi") }
func go_open(uintptr, uintptr, uintptr)           { panic("nyi") }
func go_close(uintptr, uintptr)                   { panic("nyi") }
func go_lseek(uintptr, uintptr, uintptr, uintptr) { panic("nyi") }
func go_read(uintptr, uintptr, uintptr, uintptr)  { panic("nyi") }

func go_write(fd, buf_, n_, ret_ uintptr) {
	var f *os.File
	switch fd {
	case 0:
		f = os.Stdin
	case 1:
		f = os.Stdout
	case 2:
		f = os.Stderr
	default:
		panic("file descriptors not yet implemented")
	}

	n := int32(n_)
	if n < 0 || uintptr(n) != n_ {
		panic("too many bytes to write") //FIXME
	}

	buf := make([]byte, int(n))
	for i := 0; i < int(n); i++ {
		buf[i] = *peek(buf_ + uintptr(i))
	}

	written, err := f.Write(buf)
	if err != nil {
		//FIXME do something
	}
	*((*int32)(unsafe.Pointer(ret_))) = int32(written)
}

func go__exit(uintptr) { panic("nyi") }
func go_abort()        { panic("nyi") }

type alloc struct {
	p unsafe.Pointer
	s []byte
}

var allocs = make(map[uintptr]alloc)

func go_realloc(ptr uintptr, size uint32, pnewptr uintptr) {
	var newptr uintptr = 0
	switch {
	case ptr == 0 && size == 0:
		break
	case ptr == 0: // malloc
		//TODO: if (size > int32.MAX) return 0; and set errno?
		if int(size) < 0 {
			break
		}
		buf := alloc{s: make([]byte, int(size))}
		buf.p = unsafe.Pointer(&buf.s[0])
		allocs[uintptr(buf.p)] = buf
		newptr = uintptr(buf.p)
	case size == 0: // free
		delete(allocs, ptr)
		break
	default: // realloc
		tmp := allocs[ptr].s
		go_realloc(ptr, 0, uintptr(unsafe.Pointer(&newptr)))
		go_realloc(0, size, uintptr(unsafe.Pointer(&newptr)))
		copy(allocs[newptr].s, tmp)
	}
	*((*uintptr)(unsafe.Pointer(pnewptr))) = newptr
}

var clock0 = time.Now()

func go_clock(secs_ uintptr) {
	secs := time.Since(clock0) / time.Second
	*((*int32)(unsafe.Pointer(secs_))) = int32(secs)
}
